keps/sig-network/2438-dual-stack-apiserver/README.md

```diff
@@ -230,80 +234,177 @@ scenario of wrong-single-stack pods...
-
-      - Forces the `kubernetes` Service to single-stack if there are
-        no dual-stack apiservers in etcd. Leaves the Service unchanged
-        if it is currently dual-stack and there are also dual-stack
-        apiservers in etcd.
-
-      - Syncs Endpoints the primary-IP-family EndpointSlice. If there
-        are no dual-stack apiservers in etcd then it will delete any
-        secondary-IP-family EndpointSlice.
-
-  - API server knows about `DualStackAPIServer`, the feature gate is
-    enabled, and the apiserver has a dual-stack
-    `--service-cluster-ip-range`:
-
-      - Advertises its own dual-stack endpoint IPs to etcd
+#### Current Behavior
+
+Current kube-apiservers attempt to force the `kubernetes` Service
+definition to match its expectations on startup, but do not attempt to
+reconcile it after that point. Current apiservers always expect the
+Service to be single-stack, so any time a current apiserver reconciles
+the `kubernetes` Service, it will set it to be single-stack.
+
+For Endpoints/EndpointSlice, the apiserver can run in two different
+"endpoint reconciler" modes: `master-count` (the old algorithm), and
+`lease` (the current default, and preferred mode).
+
+In `master-count` mode, each apiserver periodically checks the
+`kubernetes` `Endpoints` object, and:
+
+  - If its own IP is not listed there, it adds it.
+
+  - If the total number of IPs listed there is now greater than the
+    expected number of masters, it removes a randomly-selected IP
+    other than its own.
+
+(The changes to `Endpoints` are likewise synced to `EndpointSlice`.)
+
+This means that when an apiserver is replaced, it may take arbitrarily
+long for the `Endpoints` to settle on the correct values (since none
+of the remaining apiservers knows which IP is the one that needs to be
+removed), but once it reaches the correct values, it will stay correct
+until the set of apiservers changes again.
+
+In the `lease` mode, each apiserver periodically writes its IP to a
+non-kubernetes-API object in etcd, along with a timestamp and a lease
+time. Then it reads all of the other IPs there, and updates the
+`Endpoints` and `EndpointSlice` to contain all of the IPs whose leases
+have not expired. So in this mode, a stale IP is guaranteed to be
+removed once its lease expires, and no apiserver will ever remove a
+valid IP from the list.
+
+#### Upgrading from current apiservers to new dual-stack apiservers
+
+As of 1.22, apiservers only reconcile the `Service` object at startup,
+and only ever reconcile a single `EndpointSlice` object for the
+primary IP family. The latter behavior is clearly a bug
+([kubernetes #101070]), and it has been suggested that the former
+behavior is a bug as well. However, if we avoid fixing these bugs
+until we add dual-stack-apiserver support, then it makes upgrades from
+old/current apiservers to future dual-stack apiservers easier:
+
+  - When the first dual-stack apiserver comes up, it would switch the
```

Comment:
The paragraph you're commenting on here is about what will happen if you try to upgrade to dual-stack apiservers while you still have some apiservers running the current (1.22) code that knows nothing about dual-stack apiservers. So there's no "intention"; it's just describing what the current code will do.

In the next section ("Upgrading from new single-stack apiservers to dual-stack apiservers"), it talks about what happens when you have a cluster containing all "new" apiserver code, but where some of the apiservers think the cluster should have single-stack apiservers and some think it should have dual-stack apiservers, and the intention there is that the apiservers understand that during upgrades/downgrades/config changes they may not all be configured the same way, so all of the apiservers operate under the rule "if any apiserver has written out dual-stack endpoints, then the Service should be dual-stack".

(Though we could instead do "if any apiserver is single-stack then the Service is single-stack" rather than "if any apiserver is dual-stack then the Service is dual-stack".)